crates/wasmtime/src/runtime/externals/tag.rs

```diff
@@ -0,0 +1,44 @@
+use crate::runtime::types::TagType;
+use crate::{
+    store::{StoreData, StoreOpaque, Stored},
+    AsContext,
+};
+
+/// A WebAssembly `tag`.
+#[derive(Copy, Clone, Debug)]
+#[repr(transparent)] // here for the C API
+pub struct Tag(pub(super) Stored<crate::runtime::vm::ExportTag>);
+
+impl Tag {
+    pub(crate) unsafe fn from_wasmtime_tag(
+        mut wasmtime_export: crate::runtime::vm::ExportTag,
+        store: &mut StoreOpaque,
+    ) -> Self {
+        use wasmtime_environ::TypeTrace;
+        wasmtime_export
+            .tag
+            .canonicalize_for_runtime_usage(&mut |module_index| {
+                crate::runtime::vm::Instance::from_vmctx(wasmtime_export.vmctx, |instance| {
+                    instance.engine_type_index(module_index)
+                })
```

Comment:
As a heads-up I think this may no longer be necessary after https://github.com/bytecodealliance/wasmtime/pull/10223 landed.